src/logic/calculate.js

```diff
@@ -0,0 +1,99 @@
+import operate from './operate';
+
+const calculate = ({ total, next, operation }, buttonName) => {
+  if (buttonName.match(/[0-9]/)) {
+    return {
+      total,
+      next: next ? `${next}${buttonName}` : `${buttonName}`,
+      operation,
+    };
+  }
+
+  if (buttonName === '+/-' && next) {
+    return {
+      total,
+      next: (-1 * parseFloat(next)).toString(),
+      operation,
+    };
+  }
+
+  if (buttonName === '+/-' && !next && total) {
+    return {
+      total: (-1 * parseFloat(total)).toString(),
+      next,
+      operation,
+    };
+  }
+
+  if (buttonName.match(/[-/X+]/) && next && total) {
+    return {
+      total: (operate(parseFloat(total), parseFloat(next), buttonName)).toString(),
+      next: null,
+      operation: buttonName,
+    };
+  }
+
+  if (buttonName.match(/[-/X+]/) && !next && total) {
+    return {
+      total,
+      next,
+      operation: buttonName,
+    };
+  }
+
+  if (buttonName.match(/[-/X+]/) && next && !total) {
+    return {
+      total: next,
+      next: null,
+      operation: buttonName,
+    };
+  }
+
+  if (buttonName === '.' && next && !next.includes('.')) {
+    return {
+      total,
+      next: `${next}.`,
+      operation,
+    };
+  }
```

Comment:
- When the `buttonName` is `.` there are many more cases that you should consider.
- [ ] What if `next` is `null`, but `operation` is not `null` and the user clicks `.` ? (this should set `next` to `0.`)
- [ ] What is `next` is `null` and `total` isn't null and the user clicks `.` ? (`.` should be included in `total`)
- [ ] What if `next` exists and `.` is already included in it but user clicks on `.` again ?

Reply:
You are right

> I believe in the third case nothing will be changed (. already in next), so this case is handled by the last return,
> 
> ```
>  return {
>       total,
>       next,
>       operation,
>   };
> ```